timemachine/fe/reweighting.py

```diff
@@ -0,0 +1,309 @@
+    batched_u_1_fxn: BatchedReducedPotentialFxn,
+) -> Callable[[Params], float]:
+    r"""assuming
+    * samples from a distribution p_ref(x) \propto(exp(-u_ref(x))
+      that has good overlap with BOTH p_0(params)(x) and p_1(params)(x),
+    * evaluation (or estimates) of log_weight(x) = -u_ref(x)
+
+    construct an estimator for the free energy difference
+    f_1(params) - f_0(params)
+
+    Parameters
+    ----------
+    samples: [N,] collection
+        x ~ p_ref(x) \propto exp(-u_ref(x))
+    log_weights: [N,] array
+        log_weights[n] = -u_ref(samples[n])
+    batched_u_0_fxn
+        computes batch of endstate 0 energies at specified params
+        [u_0(x, params) for x in samples]
+    batched_u_1_fxn
+        computes batch of endstate 1 energies at specified params
+        [u_1(x, params) for x in samples]
+
+    Returns
+    -------
+    estimate_delta_f
+        computes an estimate f(params, 1) - f(params, 0) for arbitrary params
+
+        notes:
+        * estimate_delta_f(ref_params) == ref_delta_f
+        * estimate_delta_f(params) can become unreliable when
+          params is very different from ref_params
+
+    Notes
+    -----
+    * This is essentially `computePerturbedFreeEnergies` in pymbar [1], but written in a slightly more generic way.
+        (Allows the samples to come from sources other than an MBAR mixture, produces a function that can be
+        differentiated w.r.t. params if batched_u_0_fxn, batched_u_0_fxn are differentiable w.r.t. params.)
+    * Reweighting from a single reference state is used in works like ref [2] in the context of force field fitting
+    * Forming a single reference state as a mixture of several states (i.e. a constant denominator "log_weights")
+        and differentiating the numerator ("-u(samples, params)") wr.t. params
+        is used in works like ref [3] to differentiate free energy estimates w.r.t. params.
+    * Non-requirement: u_ref does not have to be of the same functional form as u_0, u_1
+
+    References
+    ----------
+    [1] pymbar implementation of computePerturbedFreeEnergies
+        https://github.com/choderalab/pymbar/blob/3c4262c490261110a7595eec37df3e2b8caeab37/pymbar/mbar.py#L1163-L1237
+    [2] Messerly RA, Razavi SM, and Shirts MR. Configuration-Sampling-Based Surrogate Models for Rapid
+        parameterization of Non-Bonded Interactions.
+        J. Chem. Theory Comput. 2018, 14, 6, 3144–3162 https://doi.org/10.1021/acs.jctc.8b00223
+    [3] Wieder et al. PyTorch implementation of differentiable reweighting in neutromeratio
+        https://github.com/choderalab/neutromeratio/blob/2abf29f03e5175a988503b5d6ceeee8ce5bfd4ad/neutromeratio/parameter_gradients.py#L246-L267
+    """
+    assert len(samples) == len(log_weights)
+
+    def f_0(params):
+        """estimate f(params, 0) - f(ref) by reweighting"""
+        log_numerator = -batched_u_0_fxn(samples, params)
+        log_importance_weights = log_numerator - log_weights
```

Comment:
I probably need to offline this with you - but I'm slightly confused as to why `log_weights` is not dependent on `params`.

Reply:
Should probably rename these arguments to be less confusing, perhaps to (`samples_ref_n`, `u_ref_n`), or similar?

https://github.com/proteneer/timemachine/blob/9f7065909e633ce8a25319d178cf7168de8d64ac/timemachine/fe/reweighting.py#L234-L237

It is an important feature of this approach that the reference state does not depend on params. We do not want to differentiate through whatever process is used to generate samples from the reference state.

When reweighting to new params, we will instead compute `w(x, params) = p_trial(x, params) / p_ref(x)`. The numerator depends on `params`, but the denominator does not.

Reply:
Name collision addressed in https://github.com/proteneer/timemachine/pull/654/commits/297f06100319b4749016aed21b459eb2024dcaa7 -- I think this is an improvement